lib/dataformat/db2sql/db2sql.3
.TH DB2SQL 3 "$Date: 2007/08/23 16:24:12 $"
.SH NAME
dbschema2sqlcreate, db2sqlinsert, db2sqldelete, db2sql_set_syncfield_name \- routines to create SQL commands from Datascope databases
.SH SYNOPSIS
.nf
#include "db2sql.h"

Tbl *dbschema2sqlcreate( Dbptr db, long flags );

long db2sqlinsert( Dbptr db, Tbl **tbl, char *(*createsync)(Dbptr db), long flags );

int db2sqldelete( Dbptr db, char *sync, Tbl **tbl, long flags );

char *db2sql_set_syncfield_name( void );

void db2sql_set_syncfield_name( char *name );

.fi
.SH DESCRIPTION
These routines create Structured Query Language (SQL) representations of Datascope databases 
and schemata. \fBdbschema2sqlcreate\fP maps a database schema from Datascope to SQL. 
\fBdb2sqlinsert\fP maps individual database rows from Datascope to SQL. 

The \fBdbschema2sqlcreate\fP function converts the schema for the input database
into a Tbl (see tbl(3)) of string values containing SQL create statements. 

Currently, the only supported value for \fIflags\fP given to \fBdbschema2sqlcreate\fP is \fIDB2SQL_OMIT_SYNC\fP.
Unless the \fIDB2SQL_OMIT_SYNC\fP flag 
is given, an extra database field is appended by \fBdbschema2sqlcreate\fP with the 
value of the macro \fIDB2SQL_SYNCFIELD_NAME_DEFAULT\fP, currently set to \fIsyncstring\fP in db2sql.h and 
with the corresponding SQL schema format given by the macro \fIDB2SQL_SYNCFIELD_SPEC\fP as 
.nf

        CHAR(40) DEFAULT '-'

.fi
This syncronization field is used to hold values of a synchronization string computed by the 
\fBdb2sqlinsert\fP function for updating changes made to the database. The default name of the 
synchronization string may be changed with the \fBdb2sql_set_syncfield_name\fP function 
described below. 

The table returned by \fBdbschema2sqlcreate\fP is newly allocated, along with all string 
fields it contains as entries, and must therefore be freed by the caller after use with a 
command like
.nf

        freetbl( tbl, free );
.fi

The \fBdb2sqlinsert\fP function generates SQL insert commands for each row in the 
specified database pointer. 

Currently, the only supported values for \fIflags\fP given to \fBdb2sqlinsert\fP are \fIDB2SQL_OMIT_SYNC\fP and 
\fIDB2SQL_USE_DATASCOPE_NULLS\fP. 

Unless the \fIDB2SQL_OMIT_SYNC\fP flag 
is used, an extra database field is computed and inserted by \fBdb2sqlinsert\fP, containing
a synchronization string that uniquely labels the contents of the input database row. The construction 
of the synchronization string for each database row is performed by the \fIcreatesync\fP function
passed to \fBdb2sqlinsert\fP.  This synchronization string can 
be used in later operations to alter or delete SQL-database rows when their corresponding Datascope database
rows change. The createsync function should be of the form
.nf

        char *\fIcreatesync\fP( Dbptr \fIdb\fP )

.fi
This function should return a newly allocated pointer to the synchronization string for the database 
row pointed to by the input database-pointer \fIdb\fP, or a null string in the case of error. The
\fBdb2sqlinsert\fP function will free this returned pointer if it is non-null. 

Unless the flag \fIDB2SQL_USE_DATASCOPE_NULLS\fP is given, \fBdb2sqlinsert\fP converts all Datascope 
null values into official SQL \fINULL\fP values. With the \fIDB2SQL_USE_DATASCOPE_NULLS\fP flag
set, Datascope Null values are inserted into the SQL tables verbatim. 

String entries added to the given table by \fBdb2sqlinsert\fP are newly allocated, and must be 
freed by the caller when no longer needed. Similarly, if the \fItbl\fP pointer given to 
\fBdb2sqlinsert\fP is NULL, \fBdb2sqlinsert\fP allocates a new table for the return results, which 
must be freed after use with a command like 
.nf

	freetbl( tbl, free );
.fi

The \fBdb2sqldelete\fP function creates a SQL delete statement for a row in the table pointed to 
by the input database pointer and described by the syncronization string given in the \fIsync\fP 
argument. The SQL command is added to the input tbl, which, as in the other functions in this library, 
is created if null and must be destroyed by the caller. 

The \fBdb2sql_get_syncfield_name\fP returns the current name of the synchronization field for the SQL 
database. This is returned as a newly allocated character string which must be freed by the caller. 

The \fBdb2sql_set_syncfield_name\fP function allows the user to specify the name of the synchronization 
field name in the SQL database, which defaults to \fIsyncstring\fP. 
.SH EXAMPLE
.in 2c
.ft CW
.nf

#include <stdlib.h>
#include "db.h"
#include "stock.h"
#include "dbmon.h"
#include "db2sql.h"

void newrow( Dbptr db, char *table, long irecord, char *sync, void *private );
void delrow( Dbptr db, char *table, char *sync, void *private );

void
newrow( Dbptr db, char *table, long irecord, char *sync, void *private )
{ 
	Tbl	**sql = (Tbl **) private;

	db2sqlinsert( db, sql, dbmon_compute_row_sync, 0L );

	return;
}

void
delrow( Dbptr db, char *table, char *sync, void *private )
{ 
	Tbl	**sql = (Tbl **) private;
	char	*sqlcmd;

	allot( char *, sqlcmd, STRSZ );

	sprintf( sqlcmd, "DELETE from `%s` WHERE %s = '%s';\\n", 
			table, DB2SQL_SYNCFIELD_NAME, sync );

	pushtbl( *sql, sqlcmd );

	return;
}

int
main(int argc, char **argv )
{
	Dbptr	db;
	char	*dbname = "./testdb/demo";
	Hook	*dbmon_hook = NULL;
	Tbl	*tables; 
	Tbl	*sql = NULL;

	if( argc > 1 ) {

		tables = strtbl( "sensor", 0 );

	} else {
		
		tables = (Tbl *) NULL;
	}

	dbopen_database( dbname, "r", &db );

	dbmon_hook = dbmon_init( db, tables, newrow, delrow, 0 );

	dbmon_update( dbmon_hook, (void *) &sql );

	dbmon_status( stdout, dbmon_hook );

	debugtbl( stdout, "\\n\\nSQL statement table:\\n", sql );

	clrtbl( sql, free );

	fprintf( stdout, "Sleeping 5 sec before checking for updates..." );
	fflush( stdout );
	sleep( 5 );
	fprintf( stdout, "awake again\\n" );

	dbmon_update( dbmon_hook, (void *) &sql );

	elog_flush( 1, 0 );

	debugtbl( stdout, "\\n\\nSQL statement table:\\n", sql );

	dbmon_close( &dbmon_hook );
	
	freetbl( sql, free ); 
}

.fi
.ft R
.in
.SH RETURN VALUES
\fBdbschema2sqlcreate\fP returns a table of SQL create statements. Both the table and the individual
statements are newly allocated, and must be freed by the caller. 

\fBdb2sqlinsert\fP returns the number of SQL commands successfully created. 

\fBdb2sql_get_syncfield_name\fP returns a newly-allocated character string which must be freed by the caller. 

.SH "SEE ALSO"
.nf
db2sql(1), perldb2sql(3P)
.fi
.SH "BUGS AND CAVEATS"
Primary keys indicating ranges in Datascope, e.g. "time::endtime", are decoupled and added to the SQL primary key 
as individual fields (e.g. PRIMARY KEY( ... time, endtime, ... )), since SQL schemata appear not to support ranges 
as primary keys. 

The SQL format emitted by \fBdb2sql\fP is currently tailored for MySQL. The option is open in the future to expand 
this to support other SQL variants. 

If no \fIcreatesync\fP function is specified to \fBdb2sqlinsert\fP, the synchronization string will be left as NULL.

MySQL SQL does not allow a default value for fields of type \fITEXT\fP, therefore none is used for fields of this type. 

If default values are being defined in the SQL version of the Datascope schema, invalid null values in the Datascope 
schema (i.e. null values not conforming to the stated format of the attribute) will prevent the corresponding table 
from being created in the SQL database. 
.SH AUTHOR
.nf
Kent Lindquist 
Lindquist Consulting, Inc.
.fi
.\" $Id: manpage_template,v 1.13 2007/08/23 16:24:12 kent Exp $


lib/dataformat/db2sql/db2sql.c

/*
 *   Copyright (c) 2009-2010 Lindquist Consulting, Inc.
 *   All rights reserved. 
 *                                                                     
 *   Written by Dr. Kent Lindquist, Lindquist Consulting, Inc. 
 *
 *   This software is licensed under the New BSD license: 
 *
 *   Redistribution and use in source and binary forms,
 *   with or without modification, are permitted provided
 *   that the following conditions are met:
 *   
 *   * Redistributions of source code must retain the above
 *   copyright notice, this list of conditions and the
 *   following disclaimer.
 *   
 *   * Redistributions in binary form must reproduce the
 *   above copyright notice, this list of conditions and
 *   the following disclaimer in the documentation and/or
 *   other materials provided with the distribution.
 *   
 *   * Neither the name of Lindquist Consulting, Inc. nor
 *   the names of its contributors may be used to endorse
 *   or promote products derived from this software without
 *   specific prior written permission.

 *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 *   WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *   WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 *   PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
 *   THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY
 *   DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 *   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 *   IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 *   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 *   USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *   POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "db.h"
#include "stock.h"
#include "db2sql.h"
#include "crc.h"

static int find_longest( void *s, void *private );
static int dbfield_isnull( Dbptr db );
static char *generate_sqltable_create( Dbptr db, long flags );
static char *generate_sqlrow_insert( Dbptr db, char *(*createsync)(Dbptr db), long flags );
static long generate_sqltable_insert( Dbptr db, Tbl **tbl, char *(*createsync)(Dbptr db), long flags );

static int
find_longest( void *s, void *longest )
{
	long	l;

	l = strlen( (char *) s );

	if( l > *((long *) longest) ) {

		*((long *) longest) = l;
	}

	return 0;
}

static int
dbfield_isnull( Dbptr db )
{
	char	*fnull = NULL;
	long	fsize;
	char	*stringval = NULL;
	int	isnull = 0;

	dbquery( db, dbFIELD_SIZE, &fsize );
	dbquery( db, dbNULL, &fnull );

	allot( char *, stringval, fsize + 2 );

	dbget( db, stringval );

	strtrim( stringval );

	if( ! strcmp( stringval, fnull ) ) {

		isnull = 1;

	} else {

		isnull = 0;
	}

	free( stringval );

	return isnull;
}

static char *
generate_sqlrow_insert( Dbptr db, char *(*createsync)(Dbptr db), long flags )
{
	void	*stk = 0;
	char	*table;
	Tbl	*fields;
	Dbvalue	dbvalue;
	char	*field;
	long	ifield;
	long	ftype;
	char	*fformat;
	char	part[STRSZ];
	char	*sync;
	char	*copy;

	if( db.record < 0 && 
	    db.record != dbSCRATCH && 
	    db.record != dbNULL ) {

		return NULL;
	}

	dbquery( db, dbTABLE_NAME, &table );
	dbquery( db, dbTABLE_FIELDS, &fields );

	pushstr( &stk, "INSERT INTO `" );
	pushstr( &stk, table );

	pushstr( &stk, "` VALUES(" );

	for( ifield = 0; ifield < maxtbl( fields ); ifield++ ) {
		
		field = gettbl( fields, ifield );

		db = dblookup( db, "", "", field, "" );

		dbquery( db, dbFIELD_TYPE, &ftype );
		dbquery( db, dbFIELD_FORMAT, &fformat );

		dbgetv( db, 0, field, &dbvalue, NULL );

		if( ifield > 0 ) {

			pushstr( &stk, ", " );
		}

		if( ! ( flags & DB2SQL_USE_DATASCOPE_NULLS ) && dbfield_isnull( db ) ) {

			pushstr( &stk, "NULL" );

			continue;
		}

		memset( part, '\0', STRSZ );

		switch( ftype ) {

		case dbSTRING:
			pushstr( &stk, "'" );
			if( strchr( dbvalue.s, '\'' ) == (char *) NULL ) {
				pushstr( &stk, dbvalue.s );
			} else {
				allot( char *, copy, 2 * strlen( dbvalue.s ) );
				strsub( dbvalue.s, "'", "\\'", copy );
				pushstr( &stk, copy );
				free( copy );
			}
			pushstr( &stk, "'" );
			break;

		case dbREAL:
		case dbTIME:
			sprintf( part, fformat, dbvalue.d );
			pushstr( &stk, part );
			break;

		case dbINTEGER:
		case dbYEARDAY:
			sprintf( part, fformat, dbvalue.i );
			pushstr( &stk, part );
			break;

		case dbDBPTR:
			sprintf( part, "%ld %ld %ld %ld", 
					dbvalue.db.database, 
					dbvalue.db.table, 
					dbvalue.db.field, 
					dbvalue.db.record );
			pushstr( &stk, part );
			break;
		}
	}

	if( ! ( flags & DB2SQL_OMIT_SYNC ) ) {

		if( createsync != (char *(*)(Dbptr db)) NULL ) {

			sync = (*createsync)( db );

		} else {

			sync = (char *) NULL;
		}

		pushstr( &stk, ", '" );

		if( sync != (char *) NULL ) {

			pushstr( &stk, sync );

		} else {

			pushstr( &stk, DB2SQL_SYNCFIELD_NULL );
		}

		pushstr( &stk, "'" );

		if( sync != (char *) NULL ) {

			free( sync );
		}
	}

	pushstr( &stk, ");\n" );

	return popstr( &stk, 1 );
}

static char *
generate_sqltable_create( Dbptr db, long flags )
{
	char	*table;
	char	part[STRSZ];
	void	*stk = 0;
	Tbl	*primary;
	Tbl	*fields;
	long	ifield;
	char	*field;
	char	field_a[STRSZ];
	char	field_b[STRSZ];
	char	*fnull;
	long	fsize;
	long	ftype;
	char	*fformat;
	int	precision;
	int	scale;
	long	longest = 0;

	if( db.table < 0 ) {

		return NULL;
	}

	dbquery( db, dbTABLE_NAME, &table );

	pushstr( &stk, "CREATE TABLE `" );
	pushstr( &stk, table );
	pushstr( &stk, "`\n  (\n" );

	dbquery( db, dbTABLE_FIELDS, &fields );

	applytbl( fields, find_longest, (void *) &longest );

	if( ! ( flags & DB2SQL_OMIT_SYNC ) && longest < strlen( Db2sql_syncfield_name ) ) {

		longest = strlen( Db2sql_syncfield_name );
	}

	for( ifield = 0; ifield < maxtbl( fields ); ifield++ ) {

		if( ifield > 0 ) {

			pushstr( &stk, ",\n" );
		}

		field = gettbl( fields, ifield );

		db = dblookup( db, "", table, field, "" );

		pushstr( &stk, "  `" );
		pushstr( &stk, field );
		pushstr( &stk, "`" );
		pushstr( &stk, spaces( longest - strlen(field) + 2 ) );

		dbquery( db, dbFIELD_SIZE, &fsize );
		dbquery( db, dbFIELD_TYPE, &ftype );
		dbquery( db, dbFIELD_FORMAT, &fformat );
		dbquery( db, dbNULL, &fnull );

		memset( part, '\0', STRSZ );

		switch( ftype ) {

		case dbSTRING:
			if( fsize < 256 ) {
				sprintf( part, "CHAR(%ld)", fsize );
			} else {
				sprintf( part, "TEXT(%ld)", fsize );
			}
			pushstr( &stk, part );
			break;

		case dbREAL:
		case dbTIME:
			if( fnull != (char *) NULL && strcontains( fnull, "[eE]", 0, 0, 0 ) ) {
				sprintf( part, "DOUBLE" );
			} else {
				sscanf( fformat, "%%%d.%d", &precision, &scale );
				sprintf( part, "DECIMAL(%d,%d)", precision, scale );
			}
			pushstr( &stk, part );
			break;

		case dbINTEGER:
		case dbYEARDAY:
			sprintf( part, "INTEGER(%ld)", fsize );
			pushstr( &stk, part );
			break;

		case dbDBPTR:
			pushstr( &stk, "CHAR(32)" );
			break;
		}

		/* Disallow defaults for SQL TEXT fields */

		if( fnull != (char *) NULL && strncmp( part, "TEXT", 4 ) ) {

			pushstr( &stk, " DEFAULT " );

			if( ftype == dbSTRING ) {
		
				pushstr( &stk, "'" );
				pushstr( &stk, fnull );
				pushstr( &stk, "'" );

			} else {

				pushstr( &stk, fnull );
			}
		}
	}

	if( ! ( flags & DB2SQL_OMIT_SYNC ) ) {

		pushstr( &stk, ",\n" );

		pushstr( &stk, "  `" );
		pushstr( &stk, Db2sql_syncfield_name );
		pushstr( &stk, "`" );
		pushstr( &stk, spaces( longest - strlen(Db2sql_syncfield_name) + 2 ) );

		pushstr( &stk, DB2SQL_SYNCFIELD_SPEC );
	}

	dbquery( db, dbPRIMARY_KEY, &primary );

	if( maxtbl( primary ) > 0 ) {

		pushstr( &stk, ",\n  PRIMARY KEY (" );

		for( ifield = 0; ifield < maxtbl( primary ); ifield++ ) {

			if( ifield > 0 ) {

				pushstr( &stk, ", " );
			}
	
			field = gettbl( primary, ifield );

			if( strcontains( field, "::", 0, 0, 0 ) ) {

				strsub( field, "::", "  ", field );

				sscanf( field, "%s %s", field_a, field_b );

				pushstr( &stk, "`" );
				pushstr( &stk, field_a );
				pushstr( &stk, "`" );

				/* end-of-range keys can be NULL in Datascope,
				   so don't use it as primary in SQL: 

				pushstr( &stk, ", `" );
				pushstr( &stk, field_b );
				pushstr( &stk, "`" );
				*/

			} else {

				pushstr( &stk, "`" );
				pushstr( &stk, field );
				pushstr( &stk, "`" );
			}
		}

		pushstr( &stk, ")" );
	}

	pushstr( &stk, "\n  );\n" );

	return popstr( &stk, 1 );
}

static long
generate_sqltable_insert( Dbptr db, Tbl **tbl, char *(*createsync)(Dbptr db), long flags ) 
{
	char	*cmd;
	long	nrecs;

	if( *tbl == (Tbl *) NULL ) {
		
		*tbl = newtbl( 0 );
	}

	dbquery( db, dbRECORD_COUNT, &nrecs ); 

	for( db.record = 0; db.record < nrecs; db.record++ ) {
			
		cmd = generate_sqlrow_insert( db, createsync, flags );

		pushtbl( *tbl, cmd );
	}

	return nrecs;
}

int
db2sqldelete( Dbptr db, char *sync, Tbl **tbl, long flags )
{
	void	*stk = 0;
	char	*table;
	char	*cmd;

	if( *tbl == (Tbl *) NULL ) {
		
		*tbl = newtbl( 0 );
	}

	dbquery( db, dbTABLE_NAME, &table );

	pushstr( &stk, "DELETE from `" );
	pushstr( &stk, table );

	pushstr( &stk, "` WHERE " );

	pushstr( &stk, "  `" );
	pushstr( &stk, Db2sql_syncfield_name );
	pushstr( &stk, "` = '" );
	pushstr( &stk, sync );
	pushstr( &stk, "'" );
	pushstr( &stk, ";\n" );

	cmd = popstr( &stk, 1 );

	pushtbl( *tbl, cmd );

	return 1;
}

long 
db2sqlinsert( Dbptr db, Tbl **tbl, char *(*createsync)(Dbptr db), long flags )
{
	char	*cmd;
	long	ncmds = 0;
	long	table_is_view = 0;
	Tbl	*tables;
	char	*table;
	long	itable;

	if( *tbl == (Tbl *) NULL ) {
		
		*tbl = newtbl( 0 );
	}

	if( db.database == dbINVALID ||
	    db.table == dbINVALID ) {

		return 0;

	} else if( db.table == dbALL ) {

		dbquery( db, dbSCHEMA_TABLES, &tables );

		for( itable = 0; itable < maxtbl( tables ); itable++ ) {

			table = gettbl( tables, itable );

			db = dblookup( db, "", table, "", "" );

			ncmds += generate_sqltable_insert( db, tbl, createsync, flags );
		}

	} else if( db.table >= 0 ) {

		dbquery( db, dbTABLE_IS_VIEW, &table_is_view );

		if( table_is_view ) {

			elog_complain( 0, "db2sqlinsert does not support non-base-table views\n" );

			return 0;

		} else {

			if( db.record == dbSCRATCH || 
			    db.record == dbNULL || 
			    db.record >= 0 ) {

				cmd = generate_sqlrow_insert( db, createsync, flags );

				pushtbl( *tbl, cmd );

				ncmds++;

			} else {

				ncmds += generate_sqltable_insert( db, tbl, createsync, flags );
			}
		}
	}

	return ncmds;
}

Tbl *
dbschema2sqlcreate( Dbptr db, long flags )
{
	Tbl	*sql;
	char	*cmd;
	Tbl	*tables;
	char	*table;
	long	itable;

	sql = newtbl( 0 );

	if( db.table >= 0 ) {
		
		cmd = generate_sqltable_create( db, flags );

		pushtbl( sql, cmd );

	} else {

		dbquery( db, dbSCHEMA_TABLES, &tables );

		for( itable = 0; itable < maxtbl( tables ); itable++ ) {

			table = gettbl( tables, itable );

			db = dblookup( db, "", table, "", "" );
			
			cmd = generate_sqltable_create( db, flags );

			pushtbl( sql, cmd );
		}
	}

	return sql;
}

void 
db2sql_set_syncfield_name( char *name )
{
	if( name != (char *) NULL ) {

		if( Db2sql_syncfield_name != (char *) NULL &&
		    strcmp( Db2sql_syncfield_name, DB2SQL_SYNCFIELD_NAME_DEFAULT ) ) {

			free( Db2sql_syncfield_name );    	
		}

		Db2sql_syncfield_name = strdup( name );
	}
}
